Kin Android/app/src/main/java/kin/unity/BackupActivity.java

```diff
@@ -0,0 +1,179 @@
+        });
+
+        backupManager.registerRestoreCallback(new RestoreCallback() {
+            @Override
+            public void onSuccess(KinClient kinClient, KinAccount kinAccount) {
+                Log.d(LOGTAG, "Got success from backup manager");
+                String accountId = _plugin.getAccountIdByAccount(kinAccount);
+                // If we already have this account, there is no need to add it again to the list.
+                // The account will still be added internally to the kin client until this bug will be fixed in the android b&r module
+                if (accountId == null)
+                {
+                    accountId = _plugin.generateUniqueId();
+                    _plugin._accounts.put(accountId, kinAccount);
+                }
+                else
+                {
+                    Log.d(LOGTAG, "Restored existing account");
+                }
+                _plugin.unitySendMessage("RestoreSucceeded", _plugin.callbackToJson(accountId, clientId));
+            }
+
+            @Override
+            public void onCancel() {
+                Log.d(LOGTAG, "Got cancel from backup manager");
+                _plugin.unitySendMessage("RestoreCanceled", _plugin.callbackToJson(null, clientId));
+            }
+
+            @Override
+            public void onFailure(BackupAndRestoreException e) {
+                Log.e(LOGTAG, "Got failure from backup manager");
+                _plugin.unitySendMessage("RestoreFailed", _plugin.exceptionToJson(e, clientId));
+            }
+        });
+
+        Log.d(LOGTAG, "Created Kin backup manager");
+        return backupManager;
+
+    }
+
+    protected void backupAccount(String accountId, String clientId)
+    {
+        try
+        {
+            backupManager = createBackupAndRestoreManager(clientId, accountId);
+            KinClient client = _plugin._clients.get(clientId);
+            KinAccount account = _plugin._accounts.get(accountId);
+
+            backupManager.backup(client, account);
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+            Log.e( LOGTAG, "backupAccount failed", e );
+            _plugin.unitySendMessage("BackupFailed", _plugin.exceptionToJson(e, managerId));
+            finish();
+        }
+    }
+
+    protected void restoreAccount(String clientId)
+    {
+        try
+        {
+            backupManager = createBackupAndRestoreManager(clientId, null);
+            KinClient client = _plugin._clients.get(clientId);
+
+            backupManager.restore(client);
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+            Log.e( LOGTAG, "restoreAccount failed", e );
+            _plugin.unitySendMessage("onRestore", _plugin.exceptionToJson(e, managerId));
+            finish();
+        }
+    }
+
+    public void releaseBackupManager()
+    {
+        try
+        {
+            backupManager.release();
+        }
+        catch (Exception e)
```

Comment:
why catch is needed here?

Reply:
I dont believe there is any reason to crash here, but we don't want to crash the Unity game in case it will, so we wrap everything and return it to c#.